R/module_teal_with_splash.R

```diff
@@ -95,15 +106,89 @@ srv_teal_with_splash <- function(id, data, modules, filter = teal_slices()) {
           )
         }
       })
-
-      if (!is.reactive(raw_data)) {
-        stop("The delayed loading module has to return a reactive object.")
-      }
       raw_data
     }
 
-    res <- srv_teal(id = "teal", modules = modules, raw_data = raw_data, filter = filter)
-    logger::log_trace("srv_teal_with_splash initialized module with data { toString(get_dataname(data))}.")
+    if (!is.reactive(raw_data)) {
+      stop("The `teal_data_module` must return a reactive expression containing a `teal_data` object.", call. = FALSE)
+    }
+
+    raw_data_checked <- reactive({
+      # custom module can return error
+      data <- tryCatch(raw_data(), error = function(e) e)
+
+      # there is an empty reactive event on init!
+      if (inherits(data, "shiny.silent.error") && identical(data$message, "")) {
+        return(NULL)
+      }
+
+      # to handle qenv.error
+      if (inherits(data, "qenv.error")) {
+        validate(
+          need(
+            FALSE,
+            paste(
+              "Error when executing `teal_data_module`:\n ",
+              data$message,
+              "\n Check your inputs or contact app developer if error persists"
+            )
+          )
+        )
+      }
+
+      # to handle module non-qenv errors
+      if (inherits(data, "error")) {
+        validate(
+          need(
+            FALSE,
+            paste0(
+              "Error when executing `teal_data_module`:\n ",
+              data$message,
+              "\n Check your inputs or contact app developer if error persists"
+            )
+          )
+        )
+      }
+
+      validate(
+        need(
+          inherits(data, "teal_data"),
+          paste(
+            "Error: `teal_data_module` did not return `teal_data` object",
+            "\n Check your inputs or contact app developer if error persists"
+          )
+        )
+      )
+
+      validate(need(teal.data::datanames(data), "Data has no datanames. Contact app developer"))
+
+
+      is_modules_ok <- check_modules_datanames(modules, teal.data::datanames(data))
+      is_filter_ok <- check_filter_datanames(filter, teal.data::datanames(data))
+
+      validate(need(isTRUE(is_modules_ok), is_modules_ok))
+
+      if (!isTRUE(is_filter_ok)) {
+        showNotification(
+          "Some filters were not applied because of incompatibility with data. Contact app developer",
+          type = "warning",
+          duration = 10
+        )
+        logger::log_warn(is_filter_ok)
+      }
+
+      raw_data()
```

Comment:
Can we not return (i.e. return `NULL`) from `raw_data_checked` / `teal_data_rv_validate `?

You might be concerned about this being re-executed on `raw_data` / `teal_data_rv()` changes (if we allow this). Please note that this reactive is used only in `output$error` so what's the point of wrapping this in reactive? Can we move this all in there? Then that problem would be gone.

Reply:
`srv_teal` requires `teal_data` object and we can't send `qenv.error` there or `simpleError`. There (srv_teal) is `observer` which waits for non-NULL value from "raw_data". When non-null value is propagated further than we have unexpected `insertUI`, and app failure.